### ISSUE TITLE
I recently found an approach to get hooks in RS3(need a driver)

### ISSUE THREAD

**Issue by RicardoMilos** [tigros/HookTools#3]
Only works on win10 x64: The driver locates address of win32kfull.sys-->NtUserGetRawInputData-->HmValidateHandle
typedef PVOID (*HMVALIDATEHANDLE)(ULONGLONG Handle, char hType);
HMVALIDATEHANDLE HmValidateHandle = nullptr;
ULONGLONG CallAddr = Pattern(win32kfullbase, "B2 12 49 8B CC E8") + 6;
HmValidateHandle = (HMVALIDATEHANDLE)((CallAddr & 0xFFFFFFFF00000000) + (DWORD)CallAddr + *(PDWORD)CallAddr + 4);

Then get list of handles HHOOK from user32.dll-->gSharedInfo->aheList-->wUniq, object of tagHOOK is (tagHOOK*)HmValidateHandle((ULONGLONG)handleHHOOK, 5);

**Comment by tigros** [comment]
Hi,
Thanks for looking into this. When you say "the driver" you mean we need a driver? Can you do it without a driver?

Thanks!

**Comment by RicardoMilos** [comment]
No, I mean my approach also requires a driver to get the information of tagHOOK. By the way, function HmValidateHandle may exists in win32k.sys or win32kbase.sys for Windows whose BuildNumber is lower than 10000. We only need to change the base address to win32k.sys or win32kbase.sys for the pattern scan to be compatible with all OS versions.

**Comment by tigros** [comment]
Have a look at my other project KLog, can you fit the functions you need in there? If you step through the code in Windows 10 it seems like just data structure differences with older windows, so ideally, if that is the problem, would be to figure out the new data structures, then no need for a driver, ofc i may be mistaken.

**Comment by RicardoMilos** [comment]
https://www.blackhat.com/docs/us-17/wednesday/us-17-Schenk-Taking-Windows-10-Kernel-Exploitation-To-The-Next-Level%E2%80%93Leveraging-Write-What-Where-Vulnerabilities-In-Creators-Update-wp.pdf This article shows how to get ulClientDelta of Desktop's heap in RS2. I also heard that in RS3, p=user32!IsMenu-->HmValidateHandle(handle, btype), ulClientDelta = *(PUINT64)(p+0x20)-(UINT64)p. Haven't tried it yet.

**Comment by tigros** [comment]
I'd sure be interested if you get hooktools working in win10, even if it means touching kprocesshacker.sys, can you sign drivers? I can't, if you can please sign KLog :) Thanks for the doc it is interesting, and good luck!

**Comment by RicardoMilos** [comment]
Sadly I don't have a digital certificate, only tested my driver on testsigning mode -_- Will try to figure out how to leak kernel objects in win10.
